### ISSUE TITLE
Abandoned?

### ISSUE THREAD

**Issue by mike2307** [jwilder/nginx-proxy#1304]
# !!!PLEASE READ!!!

## Questions

If you have a question, DO NOT SUBMIT a new issue.  Please ask the question on the Q&A Group: https://groups.google.com/forum/#!forum/nginx-proxy
**Well, isn't accessible any more...**

Is this repository abandoned? If yes: is there any successor project? I don't want to run outdated software for handling secure communication...

**Comment by buekera** [comment]
Seems like this is abandoned, yes. I have difficulties getting this working but I'm not even asking for help I think.
Did anyone make a fork and is working actively on this?

**Comment by thaJeztah** [comment]
ping @jwilder PTAL

**Comment by ehough** [comment]
FWIW I replaced `nginx-proxy` with [Traefik](https://traefik.io/). It was pretty easy to setup, met all of my requirements, and doesn't have any of the bugs I was experiencing with this app. Might be worth taking a look.

**Comment by iamjaredwalters** [comment]
@ehough How do you handle local dev with Traefik given it requires a real domain name?

**Comment by buchdag** [comment]
@jwilder as the current maintainer of [`jrcs/docker-letsencrypt-nginx-proxy-companion`](https://github.com/JrCs/docker-letsencrypt-nginx-proxy-companion), I'm getting more and more concerned by the apparent abandonment of this project and the lack of feedback on your commitment to it.

`jwilder/nginx-proxy` is not some barely used project but a 11k+ stars one with 172 millions pulls on Dockerhub.

I'm not saying that you owe the users of this project your time and/or efforts (not at all), but I think that some update on what you intend to do with it is required.

**Comment by ehough** [comment]
@iamjaredwalters I don't personally use Traefik for local dev, but perhaps you could just add an entry into your host machine's `/etc/host` to get things started?

In case it's helpful, [here's](https://pastebin.com/sJCWN6gU) the `docker-compose.yml` for my Traefik server, and [here's](https://pastebin.com/5LFZB6Ve) a sample of a backend service.

**Comment by thomasd3** [comment]
yes, it would be nice to get some statement from Jason; it is a super useful project but I would understand if he wanted to move on and do something else.. it would just be great to know what's up.

**Comment by thaJeztah** [comment]
perhaps some trusted contributors can be assigned as maintainers by @jwilder, and give merge permissions.

before doing so, it's probably good to define a clear scope of the project (feature creep is always a risk, and it's good to set clear boundaries as to what the project is intended to do, and what is out of scope for the project)

**Comment by Aschenbecher** [comment]
Well, I'm just a User , not a developer, but I see alot of pull requests here, and jwilder is not answering anymore here and at docker-gen, so why not creating a fork and put more maintainer into the fork?

**Comment by jwilder** [comment]
This project is not abandoned, but I am not actively developing it currently.  I do merge small maintenance and security fixes as quickly as I can.  Bug fixes for existing features is fair game, but I could really use help with tests, documentation and external verification of changes by others.

My hesitation with new features is mainly about ensuring backwards compatibility, keeping the project easy to use and keeping it stable.  It is amazing how many people use this project considering how it started.  It was never really intended to support the myriad of features that have been proposed.  In fact, the framework it relies on (go templates/docker-gen) is getting increasingly difficult to maintain with every new option added.  I had started to refactor it to make it easier to extend, but didn’t have time to really complete it.

I’m open to suggestions and ways I to improve things though.

**Comment by thaJeztah** [comment]
Would replacing it with (e.g) https://github.com/hairyhenderson/gomplate be a solution? (not sure if it supports all features that are needed) /cc @hairyhenderson

**Comment by hairyhenderson** [comment]
@thaJeztah thanks for the mention! gomplate could be useful from a templating perspective, and I've definitely used gomplate to generate nginx config plenty in the past, but not being that familiar with this project I'm not sure either where the gaps are...

**Comment by thomasd3** [comment]
@jwilder many of us rely on you ;)

**Comment by Aschenbecher** [comment]
> This project is not abandoned, but I am not actively developing it currently. I do merge small maintenance and security fixes as quickly as I can. Bug fixes for existing features is fair game, but I could really use help with tests, documentation and external verification of changes by others.
> 
> My hesitation with new features is mainly about ensuring backwards compatibility, keeping the project easy to use and keeping it stable. It is amazing how many people use this project considering how it started. It was never really intended to support the myriad of features that have been proposed. In fact, the framework it relies on (go templates/docker-gen) is getting increasingly difficult to maintain with every new option added. I had started to refactor it to make it easier to extend, but didn’t have time to really complete it.
> 
> I’m open to suggestions and ways to improve things though.

Thank you very much for your work and your answer! I love this project, just have problems with the not starting docker-gen, and we dont see an update there for a long time :( 
Keep on with your good work. 

Cheers

**Comment by ngoonee** [comment]
@jwilder the suggestion by @thaJeztah on adding maintainers (assuming volunteers can be found) deserves consideration. You owe no one anything beyond what you yourself are able to commit, but the sheer adoption rate of this project suggests that increasing the bus factor would be beneficial.

**Comment by tkw1536** [comment]
https://github.com/nginx-proxy/nginx-proxy/issues/1457#issuecomment-719661543

**Comment by tkw1536** [comment]
@awehrfritz regarding https://github.com/nginx-proxy/nginx-proxy/pull/1470#issuecomment-778726553:

This issue seems to be tracking the lack of active maintenance. 

There are 100+ open PRs now, I think we should try to see if someone (or a group of someones) can get push access to this repo and start merging / closing them.

**Comment by awehrfritz** [comment]
> This issue seems to be tracking the lack of active maintenance.

Agreed. This seems to be the issue. Looking at the past commits and PRs it seems that this is a one-man-show here. I can see how this may become quickly unfeasible to handle, also considering that this project started apparently as a spare time project.

> There are 100+ open PRs now, I think we should try to see if someone (or a group of someones) can get push access to this repo and start merging / closing them.

Also agreed. However, I don’t know how to do that given that the seemingly only person with admin rights (@jwilder) is unresponsive. Eventually, if there is no response within a reasonable timeframe, the only remaining option might be a fork.

**Comment by DannyBen** [comment]
It seems that [Jason](https://github.com/jwilder) has been inactive on GitHub for some time now, I suspect we might already be at the fork-to-survive junction.

Looking at the [network page](https://github.com/nginx-proxy/nginx-proxy/network), it seems there are quite a few forks that have already progressed - perhaps one of them can become the new primary?

**Comment by came88** [comment]
@buchdag, you are a member of [nginx-proxy](https://github.com/orgs/nginx-proxy/people), maybe can you help here please?

**Comment by tkw1536** [comment]
I think we should at least try to reach out to @jwilder via email or other non-github means, before trying to fork. Maybe he simply has notifications for issues disabled and isn't fully aware of this issue.

**Comment by awehrfritz** [comment]
Good point @tkw1536. We could try Twitter or LinkedIn, you can find his handles from his webpage: jasonwilder.com
Couldn’t find an email address there.

Do you want to lead that effort @tkw1536?

**Comment by tkw1536** [comment]
There is an an email address in the "maintainer" label of the dockerfile in this repo. I will try and reach out later today; I'll keep this issue up-to-date.   

If I hear nothing in a few days, I'll try twitter. Unfortunately I don't have an active LinkedIn right now, so I can't reach out via that.

**Comment by buchdag** [comment]
@came88 @tkw1536 @DannyBen @awehrfritz  I've tried to reach @jwilder through various means in the past few months, without any success.

I'll bring the issue up to the rest of the nginx-proxy org as soon as possible.

I certainly do not want to see this project cease to be maintained.

**Comment by tkw1536** [comment]
OK then I'll hold off on writing emails.

**Comment by polarathene** [comment]
I'd also like to point out that the maintenance that is done is a bit concerning. I just noticed [the following PR](https://github.com/nginx-proxy/nginx-proxy/pull/1452), this change doubled the DH params size without proper justification and enabled the usage of [`-dsaparam`](https://www.openssl.org/docs/man1.0.2/man1/dhparam.html), which [appears to be safe](https://security.stackexchange.com/a/95184/244891) with the given context, but best practice is to use the [RFC 7919 pre-defined groups with strong primes](https://wiki.mozilla.org/Security/Archive/Server_Side_TLS_4.0#Pre-defined_DHE_groups).

That said, nginx-proxy-letsencrypt-companion, still uses 2048-bit DH params (generated as well instead of using RFC 7919 groups), and provisions RSA certs with key size of 4096-bit size by default. 2048-bit is plenty secure. This default was inherited from `simp_le` before switching to `acme.sh` (which defaults to 2048-bit like certbot), but due to the [config support to change default RSA key size from `simp_le` 4096-bit](https://github.com/nginx-proxy/docker-letsencrypt-nginx-proxy-companion/pull/212), they've carried over that 4096-bit RSA default :\

---

Out of curiosity, what reasons do users have here to continue using `nginx-proxy` or a fork of it for continued maintenance? Have any of you tried Traefik or Caddy which should for most be valid alternatives?

[Caddy similar to nginx has a third-party project similar to `nginx-proxy` for dynamic config](https://github.com/lucaslorentz/caddy-docker-proxy#proxying-services-vs-containers) and [endorsed by the caddy author](https://github.com/lucaslorentz/caddy-docker-proxy/issues/130#issuecomment-617455816). Caddy doesn't need this though as it's much simpler to configure and comes with HTTPS/letsencrypt by default. Instead of editing docker env vars, you can just edit a config file for caddy.

Traefik likewise has automatic cert provisioning with letsencrypt and leverages docker labels instead of env vars. I've used both and generally prefer Caddy. I have a web server that's been running `nginx-proxy` since 2017, but with the current state of the project I'm considering just migrating to Caddy or Traefik which are closer to upstream as reputable projects with funding behind them and taking security seriously.

---

Is it worth continuing this project considering the alternatives? What use cases are there that don't migrate well to Traefik or Caddy? Here's a quick example of a basic Caddy config btw, letsencrypt is implicit:

```Caddyfile
(cors) {
  @origin header Origin {args.0}
  header @origin Access-Control-Allow-Origin "{args.0}"
  header @origin Access-Control-Request-Method GET
}

api.example.com {
  import cors example.com
  reverse_proxy localhost:9000
}

example.com {
  reverse_proxy localhost:9001
}
```

That sets up `api.example.com` and `example.com` to respond to HTTPS (or if HTTP redirect to HTTPS), automates renewals, proper OCSP stapling, etc. The api section also demonstrates adding CORS support instead of handling this at the service level. If running Caddy via a container just change `localhost` to the container/service name.

**Comment by ngoonee** [comment]
Discussing traefik or caddy is IMO off-topic here. I've used the former (still do actually), but regardless of the availability of alternatives, nginx IS a preferred solution, and nginx-proxy helps tremendously with that. Alternatives are not really relevant in a discussion about forking or maintenance.

**Comment by polarathene** [comment]
> Alternatives are not really relevant in a discussion about forking or maintenance.

Fair enough, I've hidden the comment as off-topic.

I'll be personally switching to an alternative and am still interested in use cases they don't cover, but I understand that this is not the appropriate place to discuss that further :+1:

**Comment by tkw1536** [comment]
Any news @buchdag?

**Comment by buchdag** [comment]
@tkw1536 not yet, we're having difficulties scheduling a call with our respective work agendas 😒

**Comment by buchdag** [comment]
I had a call with the nginx-proxy org last week, I'll have another one next Friday, I expect to have some firm news about how we're going to handle the maintenance of `docker-gen`, `nginx-proxy` and `letsencrypt-nginx-proxy-companion` after this call. Whatever we decide the projects won't be abandoned.

**Comment by drallgood** [comment]
Who can I reach out to if I'd be interested in helping out too?

**Comment by tkw1536** [comment]
The person I see as responsible is @buchdag.

However from the comments on this issue, I think there is some overheard just for coordinating with the existing volunteers.
Also for the record: I am happy to help out too if more help is needed.

**Comment by buchdag** [comment]
Follow up :
- the nginx-proxy org is going to contact Jason Wilder regarding ownership of `jwilder/docker-gen`. Ideally we'd like it transferred to the org so we can maintain it, but if this isn't possible we'll fork the project under the nginx-proxy org. I'll have the answer to this in ~ 1 week.
- if possible we're going to get authorisation from Jason to push to `jwilder/nginx-proxy` on Dockerhub, the same as with `jrcs/docker-letsencrypt-nginx-proxy-companion`. Re enabling the automated build on Dockerhub won't be attempted: either we get the access from Jason and it will be sufficient to automate build and push with GitHub Actions, or we don't get access and there won't be anything to do. Same timeline as above.
- regardless of what happens with the two previous points, we'd like to start building and pushing the three images under a common DockerHub organisation that we have full control over. The idea being that we'll continue to push to the "legacy" registries (`jwilder/*` and `jrcs/*`) as long as possible, but the main / active registries would be those under this DockerHub org.

The only issue with the last point is that the `nginxproxy` namespace is already taken on DockerHub 😅 

@tkw1536 if I understand correctly from #1457 you're the owner of the `nginxproxy` organisation on DockerHub ? If yes, could you contact me on my public email to discuss ownership of this namespace ?

**Comment by tkw1536** [comment]
That's great news, I am indeed the owner of the namespace. I've sent you an email regarding transferring it.

**Comment by jwilder** [comment]
I'm following up with @buchdag about being a maintainer.

**Comment by buchdag** [comment]
This point has been taken care of:

> regardless of what happens with the two previous points, we'd like to start building and pushing the three images under a common DockerHub organisation that we have full control over.

The image is now available from [`nginxproxy/nginx-proxy`](https://hub.docker.com/repository/docker/nginxproxy/nginx-proxy) on DockerHub (thanks again to @tkw1536 for transferring this namespace to us). The docs will be updated to reflect this soonish.

A lot of stuff now depends on porting the test suite to GitHub Actions, which in turn depends on having `docker-gen` fixed to work properly on GitHub Actions, with in turn again depends on wether or not we'll have to fork `docker-gen` under the `nginx-proxy` org or have the original project ownership transferred to the org from @jwilder.

**Comment by tkw1536** [comment]
The "maintainer" label in [Dockerfile](https://github.com/nginx-proxy/nginx-proxy/blob/master/Dockerfile#L45) and [Dockerfile.alpine](https://github.com/nginx-proxy/nginx-proxy/blob/master/Dockerfile.alpine#L46) should probably be updated to reflect the new maintainers.

**Comment by buchdag** [comment]
@tkw1536 will do

Sorry folks, not much update last week. This has been taken care of since my last reply :

> if possible we're going to get authorisation from Jason to push to `jwilder/nginx-proxy` on Dockerhub

So the same images are now pushed to both `jwilder/nginx-proxy` and `nginxproxy/nginx-proxy`.

I'll let you know how

> the nginx-proxy org is going to contact Jason Wilder regarding ownership of `jwilder/docker-gen`. Ideally we'd like it transferred to the org so we can maintain it, but if this isn't possible we'll fork the project under the nginx-proxy org.

goes as soon as I have update on it.

**Comment by buchdag** [comment]
Quick update:

> the nginx-proxy org is going to contact Jason Wilder regarding ownership of `jwilder/docker-gen`. Ideally we'd like it transferred to the org so we can maintain it, but if this isn't possible we'll fork the project under the nginx-proxy org.

This is still ongoing.

I'm currently blocked on docker-gen update / CI fix by https://github.com/jwilder/docker-gen/issues/349

**Comment by kamermans** [comment]
Hi all, great to see some life breathed back into this project @buchdag and @jwilder!  I don't have a ton of extra time, but I'm happy to chat if you think there is a way I might be useful with the transition and/or in the new org.  I have been running a private fork of this project for many years now and it sits in front of more than a hundred production containers in my environment.

**Comment by buchdag** [comment]
@kamermans I'm happy to see that you're still around and a strong user of nginx-proxy 👍 

I'll gladly accept help if you have some spare time to put into mulching through those ~ 550 issues and ~ 80 PR (or into contributing to **nginx-proxy/acme-companion**). I'm still not fixed on how I'll proceed but obviously some triage will be needed.

@tkw1536 you'll be welcome too by the way.

As stated by @thaJeztah a while ago
> before doing so, it's probably good to define a clear scope of the project (feature creep is always a risk, and it's good to set clear boundaries as to what the project is intended to do, and what is out of scope for the project)

I can't agree more.

@jwilder replied that
> My hesitation with new features is mainly about ensuring backwards compatibility, keeping the project easy to use and keeping it stable.

Agreed too, I don't think the goal here would be to compete with something like Traefik and answer each and everyone specific need. This project is mostly liked (imho) for its ease of use and stability, so we should try to keep it that way. There are however a few "big" features that have proved to be highly sought after by the community (the biggest one being path based routing, and It deserves a discussion of its own).

Other than that I think that some focus should be given to tests and documentation. Sure they're not the most glamorous thing but having as many feature properly tested as possible and all of them documented with a clear, easy to navigate doc is one of the pillar of a maintainable project in my eyes.

In this regard:
- I'm pretty sure some features aren't tested at all and it will take some time to find which one.
- I'm personally not fond of the one page doc format.

Another area of interest would be the mutual integration of **ningx-proxy** and the **acme-companion**. I now that we have some room for improvement here.

Do you have any feedback or suggested directions from you rather large **nginx-proxy** production workload ?

**Comment by kamermans** [comment]
I agree with your comments and I think that the scope is really important here.  There are PRs and issues for every imaginable feature, but if we actually added every imaginable feature it would probably be hard to use and impossible to maintain.

I see `nginx-proxy` as an easy-to-use ingress controller, just like the role Traefik plays in Kubernetes (which I also use).
In my environment, we have a modified `nginx-proxy` to:
1. Include Brotli by default and enable it and GZIP.
2. Increase the max upload size.
3. Mount and translate SSL certificates from our internal format to that of `nginx-proxy`.

This makes it very easy to get a web project up-and-running - we simply deploy the SSL certs and an instance of `nginx-proxy` and everything on the host has SSL, etc.

Personally, I think it makes sense to merge the LetsEncrypt companion into the main repo since HTTPS support is critical now.  Also, it's fairly easy to contribute something to the main repo that breaks the LetsEncrypt companion.  Merging these allows us to test compatibility via an integration test in the main repo.

In the past I have forked `nginx-proxy` in order to manually compile OpenSSL (and other SSL libs) in order to mitigate security risks and add support for new protocols like TLS 1.3.  We should consider tagging issues/PRs that have a security concern related to them and prioritize them (there shouldn't be very many).   [HTTP/3](https://www.nginx.com/blog/introducing-technology-preview-nginx-support-for-quic-http-3/) will present a similar problem (I'd evaluate integrating [nginx-quic](https://quic.nginx.org/readme.html) or [Cloudflare Quiche](https://github.com/cloudflare/quiche/tree/master/extras/nginx) for HTTP/3 support).

**Comment by DannyBen** [comment]
> I see nginx-proxy as an easy-to-use ingress controller.

I totally agree that feature creep should be avoided. Simplicity is a key attribute in projects like this (especially given its spotty maintenance record). I would even go as far as to say that perhaps some features or complexities should be removed, in order to simplify future development and maintenance.

You mentioned LetsEncrypt - I for one, hope that this feature stays out of the nginx-proxy repo completely. Some people have no use for it, or choose to handle SSL in different ways or different layers.

Of course, if adding it is simple, and will not complicate the codebase or its maintenance, my point might be moot - but I am assuming it is not the case.

**Comment by kamermans** [comment]
Ah, great point @DannyBen I don't use the LetsEncrypt companion at all, so my point wasn't that we should enable it by default (I don't think most people would want it anyway), just that the code should not live in a different repo because it makes it harder to test and for contributors to foresee compatibility issues.

Since @buchdag is the primary maintainer going forward, and he is also the author of the LetsEncrypt companion, I figured it would make maintenance and testing more easy.  That said, I really don't have a strong opinion either way since I don't use it.

[edit] After re-reading your comment, I think I might have missed your point.  We already have LetsEncrypt support are going to be maintaining it anyway, so there's no new development here, just removing the separation of the two repos.

**Comment by Susdevlmao** [comment]
sheeeeeeeeeeeeeesh

**Comment by AndriynomeD** [comment]
Hi, it's possible add configs to proxies additional ports? What need to do for that?
Example: I want also listen ports 5000, 9000 as additional to default VIRTUAL_PORT=80 for some containers

**Comment by buchdag** [comment]
@AndriynomeD that's not the topic of this issue. Could you please open a discussion instead ?

**Comment by psy0rz** [comment]
> Sorry for the month long delay since the last update :
> 
> > the nginx-proxy org is going to contact Jason Wilder regarding ownership of `jwilder/docker-gen`. Ideally we'd like it transferred to the org so we can maintain it, but if this isn't possible we'll fork the project under the nginx-proxy org.
> 
> The repository has been transferred and is now named [`nginx-proxy/docker-gen`](https://github.com/nginx-proxy/docker-gen)
> 
> I'll answer to the points raised by @kamermans and @DannyBen asap

great! so this scary issue can be closed now? :)

**Comment by rodrigoaguilera** [comment]
There is new maintainers but I feel the project can receive a bit more love. 
I see some folks maintaining forks with features that can be interesting for the main project. An example:
https://github.com/nginx-proxy/nginx-proxy/issues/2050

Maybe is time to welcome more maintainers trying to find the balance between keeping the project updated and the complexity of having many people taking decisions over the code.

Personally I have some features that I would like to contribute upstream to nginx-proxy but seeing that there is little chance for them to be reviewed I don't feel very motivated to tidy them up.

**Comment by buchdag** [comment]
Hi folks.

Sorry for the disappearance those past six months, it's been a rough year.

I'm getting my sh*t together and I plan to get back to the maintenance of this project (and yes it'll mean onboarding more people).
